activerecord/test/cases/persistence_test.rb

```diff
@@ -350,6 +350,22 @@ def test_becomes_includes_changed_attributes
     assert_equal %w{name}, client.changed
   end
 
+  def test_becomes_initializes_missing_attributes
+    company = Company.new(name: "GrowingCompany")
+
+    client = company.becomes(LargeClient)
+
+    assert_equal 50, client.extra_size
+  end
+
+  def test_becomes_drops_extra_attributes
+    client = LargeClient.new(name: "ShrinkingCompany")
+
+    company = client.becomes(Company)
+
+    assert_equal false, company.respond_to?(:extra_size)
+  end
```

Comment:
I'm not sure what the best solution is, but I think removing source attributes may be too surprising.  For example, I would expect `assert_equal 50, client.extra_size` to pass in this test as well.  But `client.becomes(Company)` sets `client.extra_size` to `nil`.

Reply:
Reading the documentation of `becomes` the attributes of the original class should be kept, so this should return 50

Reply:
Changing `becomes` to remove extra attributes and still reference the same data is a bit tricky so I've changed it to not remove the extra attributes: This should keep the backward compatibility and add the ability to use the extra virtual attributes introduced by the target class. Looks good to you?

If not I can work on an implementation that hides the extra attributes when needed from `ActiveModel::AttributeSet` such that we can still reference the same copy but have the effect of "removing" extra attributes.